resilience4j-retry/src/main/java/io/github/resilience4j/retry/internal/RetryImpl.java

```diff
@@ -161,9 +163,15 @@ public void onComplete() {
                     failedAfterRetryCounter.increment();
                     Throwable throwable = Option.of(lastException.get())
                         .getOrElse(lastRuntimeException.get());
+                    RuntimeException maxRetriesExceeded = new MaxRetriesExceeded(
```

Comment:
I think we should create a new exception `MaxRetriesExceededException` similar  to `CallNotPermittedException` or `BulkheadFullException`, if we really throw it as part of the Public API now.

It should also store the causing retry name as a field similar to the `CallNotPermittedException` and provide a getter method. 
I guess we also need a `writableStackTrace` config property so that users can disable the stacktrace which is sometimes filling the log files too much.

Reply:
Sure, that sounds fair.
What to do with the similarly named `MaxRetriesExceeded` ? 

Also, (I maybe have understood this wrong), but if `throwable` on the row above is not null, (eg. if attempt 2 out of 3 got an ignorable exception), we'll still send it in the `MaxRetriesExceeded` event, even though we failed because we reached maximum nbr of attempts. That seems like a bug

Reply:
I think we have to keep `MaxRetriesExceeded`  for backward compatibility :( But we shouldn't reuse the exception.

Regarding line 164:
1) Let's assume the first call fails because of an exception. And 2nd, 3rd calls fail because of the result. In that case the `RetryOnErrorEvent` would contain the last thrown exception instead of `MaxRetriesExceeded`. It seems to be a bug. You are right.

1) Let's assume the first call fails because of the result. And 2nd, 3rd calls fail because of an exception. In that case the `RetryOnErrorEvent` should contain the last thrown exception.

Reply:
Regarding point 1. We wouldn't even reach the `onComplete` method, unless the last retry attempt return a non-exceptional result. Sooo, I'm not sure what the `lastRuntimeException` is supposed to mean. But let's keep it as-is.

I agree on your last statement, let's make it so to be less confusing.